src/server/process_request.c

```diff
@@ -261,6 +261,98 @@ req_authenticate(conn_t *conn, struct batch_request *request)
 	reply_ack(request);
 }
 
+#ifndef PBS_MOM
+static void
+req_register_sched(conn_t *conn, struct batch_request *preq)
+{
+	pbs_sched *sched;
+	conn_t *pconn;
+	int rc;
+	char *user = pbs_conf.pbs_daemon_service_user ? pbs_conf.pbs_daemon_service_user : pbs_current_user;
+
+	if ((conn->cn_authen & PBS_NET_CONN_AUTHENTICATED) == 0 || strcmp(conn->cn_username, user) != 0) {
+		rc = PBSE_PERM;
+		goto rerr;
+	}
+	if (preq->rq_ind.rq_rsched.rq_conn_type == NULL || preq->rq_ind.rq_rsched.rq_name == NULL) {
+		rc = PBSE_IVALREQ;
+		goto rerr;
+	}
+	sched = find_sched(preq->rq_ind.rq_rsched.rq_name);
+	if (sched == NULL) {
+		rc = PBSE_UNKSCHED;
+		goto rerr;
+	}
+	if (sched->sc_conn_addr != conn->cn_addr) {
+		rc = PBSE_BADHOST;
+		goto rerr;
+	}
+	if (sched->sc_primary_conn != -1 || sched->sc_secondary_conn != -1) {
+		rc = PBSE_SCHEDCONNECTED;
+		goto rerr;
+	}
+	if (strcmp(preq->rq_ind.rq_rsched.rq_conn_type, "primary") == 0) {
+		sched->sc_tmp_primary_conn = conn->cn_sock;
+		reply_ack(preq);
+		return;
+	} else if (strcmp(preq->rq_ind.rq_rsched.rq_conn_type, "secondary") == 0) {
+		if (sched->sc_tmp_primary_conn == -1) {
+			rc = PBSE_IVALREQ;
+			goto rerr;
+		}
+		pconn = get_conn(sched->sc_tmp_primary_conn);
+		if (!pconn) {
+			rc = PBSE_INTERNAL;
+			goto rerr;
+		}
+	} else {
+		rc = PBSE_IVALREQ;
+		goto rerr;
+	}
+	if (pconn->cn_sock == conn->cn_sock) {
+		rc = PBSE_IVALREQ;
+		goto rerr;
+	}
+	if ((pconn->cn_authen & PBS_NET_CONN_AUTHENTICATED) == 0 || strcmp(pconn->cn_physhost, conn->cn_physhost) || pconn->cn_addr != conn->cn_addr) {
+		rc = PBSE_PERM;
+		goto rerr;
+	}
+	sched->sc_tmp_primary_conn = -1;
+	sched->sc_primary_conn = pconn->cn_sock;
+	sched->sc_secondary_conn = conn->cn_sock;
+	conn->cn_authen |= PBS_NET_CONN_FROM_PRIVIL | PBS_NET_CONN_NOTIMEOUT;
+	pconn->cn_authen |= PBS_NET_CONN_FROM_PRIVIL | PBS_NET_CONN_NOTIMEOUT;
+	pconn->cn_origin = CONN_SCHED_PRIMARY;
+	conn->cn_origin = CONN_SCHED_SECONDARY;
+	net_add_close_func(conn->cn_sock, scheduler_close);
+	net_add_close_func(pconn->cn_sock, scheduler_close);
+	if (!set_conn_as_priority(pconn)) {
+		log_eventf(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SCHED, LOG_ERR, sched->sc_name, "Failed to set primary connection as priority connection");
+		rc = PBSE_INTERNAL;
+		goto rerr;
+	}
+	if (!set_conn_as_priority(conn)) {
+		log_eventf(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SCHED, LOG_ERR, sched->sc_name, "Failed to set secondary connection as priority connection");
+		rc = PBSE_INTERNAL;
+		goto rerr;
+	}
+
+	reply_ack(preq);
+	log_eventf(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SCHED, LOG_ERR, sched->sc_name, "scheduler connected");
+	/*
+	 * scheduler (re-)connected, ask it to configure itself
+	 *
+	 * this must come after above reply_ack
+	 */
+	send_sched_cmd(sched, SCH_CONFIGURE, NULL);
+	return;
+
+rerr:
+	req_reject(rc, 0, preq);
+	close_client(preq->rq_conn);
```

Comment:
You shouldn't need to close the connection after a req_reject().

Reply:
@bhroam req_reject() doesn't close connection, we have to explicitly call close connection after it.

Reply:
I guess what confused me is that this is the first time I've seen a req_reject() followed by a close_client().  Although I think the req_reject() frees the preq.  If that is the case, you'll have to save the rq_conn to close.

Reply:
Yes, I have handled saving preq->rq_conn in last commit.